`odo dev --debug` + fix sync

feloy opened the pull request:
**What type of PR is this:**

/kind feature

**What does this PR do / why we need it:**

Adds the `debug` flag to the `odo dev` command

The user needs to add the debug port as an endpoint to the devfile component.

Fix ignores files during watch + sync

**Which issue(s) this PR fixes:**

Fixes #5583
Fixes #4389

**PR acceptance criteria:**

- [ ] Unit test 

- [x] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**

cdrage commented on pkg/odo/cli/dev/dev.go:
```diff
@@ -292,7 +294,8 @@ It forwards endpoints with exposure values 'public' or 'internal' to a port on l
 	}
-	devCmd.Flags().BoolVarP(&o.randomPorts, "random-ports", "f", false, "Assign random ports to redirected ports")
+	devCmd.Flags().BoolVar(&o.randomPortsFlag, "random-ports", false, "Assign random ports to redirected ports")
+	devCmd.Flags().BoolVar(&o.debugFlag, "debug", false, "Execute the debug command to start the component inside container")
```
Bit of an odd sentence. Maybe: `Excecute the debug command within the component` ?

cdrage commented on pkg/watch/watch.go:
```diff
@@ -83,7 +88,7 @@ type cleanupFunc func(devfileObj parser.DevfileObj, out io.Writer) error
 // path is the path of the file or the directory
 // ignores contains the glob rules for matching
-func addRecursiveWatch(watcher *fsnotify.Watcher, path string, ignores []string) error {
+func addRecursiveWatch(watcher *fsnotify.Watcher, topPath string, path string, ignores []string) error {
```
unsure if topPath is a spelling error or not either? should it be toPath?

feloy replied:
I changed to `rootPath` to avoid confusion, and documented it

cdrage reviewed (commented):
Super small edits, otherwise, the code LGTM / works well for me!

rm3l reviewed (commented):
Maybe I missed something while reading the issues this PR is supposed to fix (both https://github.com/redhat-developer/odo/issues/5583 and https://github.com/redhat-developer/odo/issues/4389), but it looks like they are not related (although https://github.com/redhat-developer/odo/issues/4389#issuecomment-1098863359 talks about fixing the issue for `odo dev` as well).
Wouldn't it be better to have 2 separate PRs instead, to make it easier to review both (and also read the Git history)?

feloy commented on pkg/watch/watch.go:
```diff
@@ -252,11 +275,11 @@ func evaluateFileChanges(events []fsnotify.Event, fileIgnores []string, watcher
-		matched, globErr := dfutil.IsGlobExpMatch(event.Name, fileIgnores)
-		klog.V(4).Infof("Matching %s with %s. Matched %v, err: %v", event.Name, fileIgnores, matched, globErr)
-		if globErr != nil {
-			watchError = fmt.Errorf("unable to watch changes: %w", globErr)
+		rel, err := filepath.Rel(path, event.Name)
+		if err != nil {
+			watchError = fmt.Errorf("unable to get relative path of %q on %q", event.Name, path)
```
I can see it used in 
```
if e := addRecursiveWatch(watcher, path, event.Name, fileIgnores); e != nil && watchError == nil {
```

feloy commented:
> Maybe I missed something while reading the issues this PR is supposed to fix (both #5583 and #4389), but it looks like they are not related (although [#4389 (comment)](https://github.com/redhat-developer/odo/issues/4389#issuecomment-1098863359) talks about fixing the issue for `odo dev` as well). Wouldn't it be better to have 2 separate PRs instead, to make it easier to review both (and also read the Git history)?

I fixed #4389 in this PR because `odo dev --debug` is quite unusable without it.

But it will be simpler to review and for history if I split them, I'll make a separate PR for #4389

feloy commented:
Closed in favor of #5690 + #5689
